utils.py

```diff
@@ -345,25 +348,36 @@ def get_runmode():
         return 'script'
 
 
-def query_packages(packages):
+def query_packages(mode, packages):
     if config.SKIP_DEPENDENCIES:
         return
 
     missing_packages = []
+    bad_packages = []
 
     for p in packages:
         command = f"{config.PACKAGE_MANAGER_COMMAND_QUERY}{p}"
         logging.debug(f"pkg query command: \"{command}\"")
         result = subprocess.run(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell=True, text=True)
         logging.debug(f"pkg query result: {result.returncode}")
-        if result.returncode != 0:
+        if result.returncode != 0 and mode == "install":
             missing_packages.append(p)
-
-    msg = 'None'
-    if missing_packages:
-        msg = f"Missing packages: {' '.join(missing_packages)}"
-    logging.info(f"Missing packages: {msg}")
-    return missing_packages
+        elif result.returncode == 0 and mode == "remove":
+            bad_packages.append(p)
+        elif mode != "install" and mode != "remove":
+            logging.error("query_package: Incorrect value for 'mode'.")
+
+    if mode == "install":
+        msg = 'None'
+        if missing_packages:
+            msg = f"Missing packages: {' '.join(missing_packages)}"
+        logging.info(f"Missing packages: {msg}")
+        return missing_packages
+    if mode == "remove":
+        msg = 'None'
+        if bad_packages:
+            msg = f"Bad packages: {' '.join(bad_packages)}"
+        logging.info(f"Bad packages: {msg}")
```

Comment:
Minor detail: You can define `msg = 'None'` once just before `if mode == "install":` rather than doing it twice (once after each `if mode ==` statement).